Runtime: expose demangle() in Runtime module

ktoso opened the pull request:
We should expose the demangle functionality; It's been widely used by calling into internal _swift_demangle and instead we should offer a real API. It's also already used in the Runtime module already when forming backtraces.

[Previous discussions](https://forums.swift.org/t/demangle-function/25416/15) happened between 2019 and 2024, and just never materialized in a complete implementation and proposal.

Right now, even more tools are in need of this API, as we are building continious profiling solutions etc, so it is a good time to revisit this topic.

This PR is roughly based off @Azoy's
https://github.com/swiftlang/swift/pull/25314/files#diff-fd379a721cc9a1c9ef6486eae713e945da842b42170d4d069029a57334371eba from 2019, however it brings it over to the new Runtime module which is a great place to put this functionality - even Backtrace had to recently reinvent calling the demangling infra in this module.

Pending SE review, [proposal here](https://github.com/swiftlang/swift-evolution/pull/2989).

cc @azoy @al45tair

ktoso committed:
move demangle() func to UTF8Span and OutputSpan

compnerd commented on include/swift/Runtime/Runtime.h:
```diff
@@ -0,0 +1,55 @@
+#include <sys/wait.h>
+
+#include <signal.h>
+#endif // defined(__linux__)
```
Nit: ifdef instead of defined (I do like the defined form myself). Switch either side to be homogeneous.

ktoso commented on stdlib/public/runtime/Demangle.cpp:
```diff
@@ -1033,3 +1039,93 @@ char *swift_demangle(const char *mangledName,
+                            size_t *outputBufferSize,
+                            uint32_t flags) {
+      if (flags > 1) {
+        swift::fatalError(0, "Only 'flags' value of '0' and '1' is currently supported.");
```
Yeah, let's do that -- the fatalError we do here is how `swift_demangle` behaves today, but it never was public API; Making the public entry point nicer to use/evolve then 👍

ktoso commented on stdlib/public/runtime/Demangle.cpp:
```diff
@@ -1033,3 +1039,93 @@ char *swift_demangle(const char *mangledName,
+      llvm::StringRef name = llvm::StringRef(mangledName, mangledNameLength);
+
+      // You must provide buffer size if you're providing your own output buffer
+      if (outputBuffer && !outputBufferSize) {
```
True, keeping the null return

ktoso commented on test/Concurrency/Runtime/RuntimeDemangle.swift:
```diff
@@ -0,0 +1,144 @@
+
+if #available(SwiftStdlib 6.3, *) {
+  DemangleTests.test("basic string return API") {
+    // First, test that we get back the mangled name with invalid input.
```
thx, updated and checked other comments; this used to be the behavior earlier

ktoso committed:
add tests for C++ symbol demangling

ktoso commented:
Thanks for the review! Adjusted things and fixed up the \0 story a bit, need to determine if we keep the old backtrace entrypoint at all.

I also added tests for the C++ mangling.

ktoso commented:
@swift-ci please smoke test

mikeash commented on stdlib/public/RuntimeModule/modules/Runtime/Runtime.h:
```diff
@@ -33,11 +33,29 @@ void swift_reportWarning(uint32_t flags, const char *message);
 
 // Demangle the given raw name (supports Swift and C++)
+//
+// The demangled string IS a null-terminated c-string.
```
Could we return a NUL terminated string from the C interface, but provide a span which doesn't cover the terminator? Slightly wasteful, but convenient.

ktoso replied:
Yeah maybe that way around would be easier

ktoso committed:
Update stdlib/public/RuntimeModule/Mangling.swift

Co-authored-by: Alastair Houghton <alastair@alastairs-place.net>

ktoso committed:
remove C++ demangling support from demangle()

ktoso committed:
add codepoint truncation protections to demangle()

rjmccall commented on include/swift/Runtime/Runtime.h:
```diff
@@ -0,0 +1,55 @@
+  size_t mangledNameLength,
+  char *outputBuffer,
+  size_t *outputBufferSize,
+  uint32_t flags
```
Please make this a `size_t`; integer types smaller than the register size raise all sorts of ABI questions.

ktoso replied:
Ok will do

ktoso committed:
make flags size_t

ktoso committed:
make methods always throwing rather than returning result

minor fixup in comments

minor docs fixes

ktoso commented:
@swift-ci please smoke test

ktoso commented:
@swift-ci please smoke test

ktoso committed:
stop sharing impl with backtrace, that one does demangle C++ as well

ktoso committed:
remove option in demangle impl which was only here for backtracer

Discard changes to stdlib/public/RuntimeModule/SymbolicatedBacktrace.swift

Discard changes to include/swift/Runtime/Backtrace.h

ktoso committed:
rework the runtime functions, too messy to keep the different modes in
single func;

ktoso commented:
@swift-ci please smoke test

ktoso committed:
handle malloc failures gracefully, better safe than sorry

ktoso committed:
utf8 codepoint needs lookahead

ktoso committed:
some mangling schemes are not supported on linux

ktoso commented:
@swift-ci please smoke test macOS

ktoso commented:
Mac failure was subprocess heh:

```
/Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/swift/stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift:455:10: error: 'nil' is incompatible with return type '_stdlib_posix_spawn_file_actions_t' (aka 'UnsafeMutableRawPointer')
20:35:41  453 | internal func _make_posix_spawn_file_actions_t()
20:35:41  454 |   -> _stdlib_posix_spawn_file_actions_t {
20:35:41  455 |   return nil
20:35:41      |          `- error: 'nil' is incompatible with return type '_stdlib_posix_spawn_file_actions_t' (aka 'UnsafeMutableRawPointer')
20:35:41  456 | }
20:35:41  457 | #endif
```

ktoso commented:
@swift-ci please smoke test

ktoso committed:
Discard changes to stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift

ktoso commented:
@swift-ci please smoke test

ktoso committed:
add added symbols to abi tests

ktoso commented:
@swift-ci please smoke test

ktoso committed:
remove not used include

ktoso committed:
Revert "remove not used include"

This reverts commit 434f84d9e611f9928f0e619430304878a88fbacf.
